File: src/01.array/08.search/01.303.go
package arraySearch

type NumArray struct {
	sums []int
}

func Constructor(nums []int) NumArray {
	sums := make([]int, len(nums)+1)

	for i, v := range nums {
		sums[i+1] = sums[i] + v
	}

	return NumArray{
		sums: sums,
	}
}

func (n *NumArray) SumRange(left int, right int) int {
	return n.sums[right+1] - n.sums[left]
}

File: src/01.array/08.search/01.303.ts
export class NumArray {
    sums: number[];
    constructor(nums: number[]) {
        this.sums = [0];
        // 提前计算 nums 的前缀和
        nums.forEach((item, index) => {
            this.sums[index + 1] = this.sums[index] + item;
        });
    }

    sumRange(left: number, right: number): number {
        return this.sums[right + 1] - this.sums[left];
    }
}
